Automattic/wp-calypso
# Simplify the address normalization logic

I put this in a separate PR because it's bigger than a simple "extract to function" refactor. This PR has the `dummy Customs form` branch as a target because that's the branch that prompted this refactor and it would have conflicts if I simply created it from `master`. This doesn't need to be merged in any particular order, since it's not a blocker for the international labels feature. It can be cherry-picked after international labels is released.

During the review of the international labels PR, @v18 noticed code that's repeated often: https://github.com/Automattic/wp-calypso/pull/24975#discussion_r195495224

This logic or some variation of it is repeated a lot of times in our codebase, and it's easy to make a mistake: `addressValues = address.isNormalized && address.selectNormalized ? address.normalizedValues : address.values`

https://github.com/Automattic/wp-calypso/commit/4a55fc5d4c9466af0da22687a9b2c9f16f46610c extracts that logic into its own function, with some JSDoc so hopefully it's easier to read.

To make sure all the usages of that pattern could be substituted, I simplified the `normalize-address.js` logic in https://github.com/Automattic/wp-calypso/commit/c5b9e15293c38f4683520bfa74c7c7486090deb1. Before it was over-complicated and called multiple Redux actions on completion, whereas now, when the normalization finishes, only a single action is dispatched so the Redux tree is always in a consistent state.

To test:
- Go to an order.
- Change the address in a minor way so a new address is suggested after validation.
- Select the original address when prompted for a suggestion. Select the suggested address back. Click the "edit original address" link.
- Change the address to something clearly invalid, and validate the address. After validation fails, click the button to use the entered address anyway.
- Check that, during all those steps, the summary of the Address step shows the address you expect (suggested or original, depending on what you chose).
